### PR TITLE
added "aud" property support for DCR API

### PR DESCRIPTION
**Pull request by Yasasr1**
Related to issue - [Improve DCR API to support custom validations and attributes mentioned in the OIDC specification](https://github.com/wso2/product-is/issues/10652)

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 1]
```diff
@@ -387,6 +388,10 @@ private OAuthConsumerAppDTO createOAuthApp(ApplicationRegistrationRequest regist
         oAuthConsumerApp.setTokenType(registrationRequest.getTokenType());
         oAuthConsumerApp.setBackChannelLogoutUrl(
                 validateBackchannelLogoutURI(registrationRequest.getBackchannelLogoutUri()));
+        if (!registrationRequest.getAudiences().isEmpty()) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:391]: Shall we use CollectionUtils.isEmpty here?

File: components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/RegistrationRequestDTO.java [hunk 2]
```diff
@@ -51,6 +52,12 @@ public void setClientName(String clientName) {
     this.clientName = clientName;
   }
 
+  @ApiModelProperty
+  @JsonProperty("aud")
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/RegistrationRequestDTO.java:56]: the newly introduced needs to be added as a property of Registration Request Object [1] of the swagger API definition for the endpoint.

[1] https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/main/resources/api.identity.oauth.dcr.endpoint.yaml#L264

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java (hunk 1 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java]: Fixed

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 3]
```diff
@@ -387,6 +389,25 @@ private OAuthConsumerAppDTO createOAuthApp(ApplicationRegistrationRequest regist
...
+            oAuthConsumerApp.setAudiences(registrationRequest.getAudiences().toArray(new String[0]));
+        }
+
+        if (registrationRequest.getIdTokenEncryptionAlgorithm() != null) {
+            oAuthConsumerApp.setIdTokenEncryptionEnabled(true);
+            oAuthConsumerApp.setIdTokenEncryptionAlgorithm(registrationRequest.getIdTokenEncryptionAlgorithm());
+            if (registrationRequest.getIdTokenEncryptionMethod() != null) {
+                oAuthConsumerApp.setIdTokenEncryptionMethod(registrationRequest.getIdTokenEncryptionMethod());
+            } else {
+                oAuthConsumerApp.setIdTokenEncryptionMethod("A128CBC-HS256");
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:403]: isn't there a util method to get default id token encryption method e.g. from 
OAuth2Utl?

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 4]
```diff
@@ -387,6 +389,25 @@ private OAuthConsumerAppDTO createOAuthApp(ApplicationRegistrationRequest regist
...
+            }
+        } else {
+            if (registrationRequest.getIdTokenEncryptionMethod() != null) {
+                throw  DCRMUtils.generateClientException(DCRMConstants.ErrorMessages.BAD_REQUEST_INSUFFICIENT_DATA,
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:407]: formatting issue after 'throw'

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java (hunk 3 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:403]: There is a method to get default value but it's set to "A128GCM"
[https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java#L186](url)

Comment (emswbandara) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:403]: we will need to fix it and use that because encryption method default value is configurable from identity.xml

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 5]
```diff
@@ -424,6 +509,14 @@ private OAuthConsumerAppDTO createOAuthApp(ApplicationRegistrationRequest regist
         return createdApp;
     }
 
+    private boolean validateTokenEndpointAuthMethod(String tokenEndpointAuthMethod) {
+        if (Arrays.asList(TOKEN_ENDPOINT_AUTH_METHODS).contains(tokenEndpointAuthMethod)) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:513]: let's simplify this as `return Arrays.asList(TOKEN_ENDPOINT_AUTH_METHODS).contains(tokenEndpointAuthMethod)`

File: components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMServiceTest.java [hunk 6]
```diff
@@ -558,6 +560,86 @@ public void registerApplicationTestWithFailedToUpdateSP(List<String> redirectUri
...
+        applicationRegistrationRequest.setRedirectUris(redirectUri);
+        applicationRegistrationRequest.setTokenEndpointAuthMethod(dummyInvalidTokenEndpointAuthMethod);
+
+        String grantType = StringUtils.join(applicationRegistrationRequest.getGrantTypes(), " ");
+        ServiceProvider serviceProvider = new ServiceProvider();
+        DCRDataHolder dcrDataHolder = DCRDataHolder.getInstance();
+        dcrDataHolder.setApplicationManagementService(mockApplicationManagementService);
+        when(mockApplicationManagementService.getServiceProvider(dummyClientName, dummyTenantDomain)).thenReturn
+                (null, serviceProvider);
+
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth.dcr/src/test/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMServiceTest.java:583]: remove unnecessary new lines

File: components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/main/resources/api.identity.oauth.dcr.endpoint.yaml [hunk 7]
```diff
@@ -107,7 +107,7 @@ paths:
         "redirect_uris":["https://client.example.org/callback"],
         "client_name":"application_owner_application_1",
         "grant_types": ["password"]}
-      operationId: getApplicationByName
+      operationId: getApplicationWithName
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/main/resources/api.identity.oauth.dcr.endpoint.yaml:110]: what is the reason for this change?

File: components/org.wso2.carbon.identity.oauth.dcr/pom.xml [hunk 8]
```diff
@@ -128,6 +128,8 @@
 
                             org.apache.commons.logging; version="${commons-logging.osgi.version.range}",
                             org.apache.commons.lang; version="${commons-lang.wso2.osgi.version.range}",
+                            org.apache.commons.collections; version="${commons-collections.wso2.osgi.version.range}",
+
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth.dcr/pom.xml:132]: remove unnessary new lines

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 9]
```diff
@@ -234,6 +238,44 @@ public Application updateApplication(ApplicationUpdateRequest updateRequest, Str
                 String backChannelLogoutUri = validateBackchannelLogoutURI(updateRequest.getBackchannelLogoutUri());
                 appDTO.setBackChannelLogoutUrl(backChannelLogoutUri);
             }
+            if (CollectionUtils.isNotEmpty(updateRequest.getAud())) {
+                appDTO.setAudiences(updateRequest.getAud().toArray(new String[0]));
+            }
+            if (updateRequest.getTokenEndpointAuthMethod() != null) {
+                if (validateTokenEndpointAuthMethod(updateRequest.getTokenEndpointAuthMethod())) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:245]: can't we combine the 2 if conditons and get rid of nested ifs

Comment (emswbandara) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:245]: otherwise it is better to add the null check inside validateTokenEndpointAuthMethod function

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java (hunk 5 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java]: Added null check of tokenEndpointAuthMethod to this function

File: components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/main/resources/api.identity.oauth.dcr.endpoint.yaml (hunk 7 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.oauth.dcr.endpoint/src/main/resources/api.identity.oauth.dcr.endpoint.yaml]: Previously this method was named as `getApplicationWithName` in RegisterApi class. But yaml file had the name `getApplicationByName` so auto generated RegisterApi class had that function as `getApplicationByName`. So i changed the yaml file to get the original method name `getApplicationWithName`.

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java (hunk 9 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java]: Added the null check to validateTokenEndpointAuthMethod

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 10]
```diff
@@ -69,6 +71,8 @@
     private static final String OAUTH_VERSION = "OAuth-2.0";
     private static final String GRANT_TYPE_SEPARATOR = " ";
     private static Pattern clientIdRegexPattern = null;
+    private static final String[] TOKEN_ENDPOINT_AUTH_METHODS = {"none", "client_secret_post",
+            "client_secret_basic", "private_key_jwt"};
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:75]: shouldn't we have `client_secret_jwt` here ?

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 11]
```diff
@@ -234,6 +238,39 @@ public Application updateApplication(ApplicationUpdateRequest updateRequest, Str
...
             }
+            if (CollectionUtils.isNotEmpty(updateRequest.getAud())) {
+                appDTO.setAudiences(updateRequest.getAud().toArray(new String[0]));
+            }
+
+            if (validateTokenEndpointAuthMethod(updateRequest.getTokenEndpointAuthMethod())) {
+                appDTO.setTokenEndpointAuthMethod(updateRequest.getTokenEndpointAuthMethod());
+            }
+
+            if (updateRequest.getIdTokenEncryptionAlgorithm() != null ||
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:249]: shall we use `StringUtils.isNotBlank` here

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 12]
```diff
@@ -234,6 +238,39 @@ public Application updateApplication(ApplicationUpdateRequest updateRequest, Str
...
+                }
+                if (appDTO.getIdTokenEncryptionAlgorithm() != null &&
+                        appDTO.getIdTokenEncryptionMethod() == null) {
+                    appDTO.setIdTokenEncryptionMethod(OAuthServerConfiguration.getInstance().
+                            getDefaultIdTokenEncryptionMethod());
+                }
+
+            }
+
+            if (updateRequest.getSoftwareId() != null) {
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:271]: refer https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1560/files#r607495816

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 13]
```diff
@@ -424,6 +499,20 @@ private OAuthConsumerAppDTO createOAuthApp(ApplicationRegistrationRequest regist
         return createdApp;
     }
 
+    private boolean validateTokenEndpointAuthMethod(String tokenEndpointAuthMethod) throws DCRMException {
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:502]: a better method name would be `isValidTokenEndpointAuthMethod`

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java (hunk 11 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:243]: In this implementation if someone send `id_token_encrypted_response_alg` as a blank value (" ") in the update request, IS will not update the value and throw an error. But if we use `StringUtils.isNotBlank` IS will not update the value but will return a 200 response. Is that ok ?

Comment (Yasasr1) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:243]: because `StringUtils.isBlank(" ") = true`

Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:243]: thanks for the explanation.
adding for reference: `id_token_encrypted_response_alg = " "` will be validated at https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java#L1392

Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:243]: @mefarazath the ideal way to handle `id_token_encrypted_response_alg = " "` would be to throw an error instead of ignoring it right (just wanted to confirm the recommended approach)?

Comment (mefarazath) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:243]: Yes. Throwing a Bad request error would be correct way since the user sends an invalid value

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 14]
```diff
@@ -234,6 +236,35 @@ public Application updateApplication(ApplicationUpdateRequest updateRequest, Str
...
+                if (appDTO.getIdTokenEncryptionAlgorithm() == null &&
+                        appDTO.getIdTokenEncryptionMethod() != null) {
+                    throw DCRMUtils.generateClientException(DCRMConstants.
+                            ErrorMessages.BAD_REQUEST_INSUFFICIENT_DATA, null);
+                }
+                if (appDTO.getIdTokenEncryptionAlgorithm() != null &&
+                        appDTO.getIdTokenEncryptionMethod() == null) {
+                    appDTO.setIdTokenEncryptionMethod(OAuthServerConfiguration.getInstance().
+                            getDefaultIdTokenEncryptionMethod());
+                }
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:261]: What is the reason for setting the default value in second if condition and throwing the error in first if condition?

Is it okay to send an IdTokenEncryptionMethod without a IdTokenEncryptionAlgorithm and not the other way?

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 15]
```diff
@@ -387,7 +428,28 @@ private OAuthConsumerAppDTO createOAuthApp(ApplicationRegistrationRequest regist
         oAuthConsumerApp.setTokenType(registrationRequest.getTokenType());
         oAuthConsumerApp.setBackChannelLogoutUrl(
                 validateBackchannelLogoutURI(registrationRequest.getBackchannelLogoutUri()));
+        if (CollectionUtils.isNotEmpty(registrationRequest.getAudiences())) {
+            OAuth2ServiceComponentHolder.setAudienceEnabled(true);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:432]: Why do we have to set this?

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java (hunk 14 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:261]: According to specs, If user sends an `idTokenEncryptionAlgorithm` without `idTokenEncryptionMethod` we should set `idTokenEncryptionMethod` to the default value. If user send `idTokenEncryptionMethod` without `idTokenEncryptionAlgorithm` we should throw an error.
[https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata](https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata)

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java (hunk 15 above)
Comment (Yasasr1) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:432]: This needs to be set to true if there are any audience values for a service provider. there is also a checkbox in management console to reflect this. 
image

Comment (mefarazath) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:432]: I don't think we need to set this here.
This would probably be a check at startup to see if relevant tables are available to store audiences. (Anyways this is not needed now)

@emswbandara any ideas why we still have this check?

Comment (emswbandara) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:432]: yes we can get rid of this check now. Shall we create a seperate git issue to track it?

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java (hunk 14 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:261]: IMO it is better if we can include this information as a comment in the code.

**Comment by mefarazath** [comment]
Closing the PR since the improvements added are not immediately required. PR can be reopen when we revisit the original issue
